main/src/ca/uwaterloo/flix/language/phase/unification/FastSetUnification.scala

```diff
@@ -0,0 +1,338 @@
+            varsNeg.iterator.map(_.x) ++
+            other.iterator.flatMap(_.variables)
+        )
+    }
+
+    /** `true` if `this` contains neither [[Var]] nor [[Cst]]. */
+    final def ground: Boolean = this match {
+      case Univ => true
+      case Empty => true
+      case Cst(_) => false
+      case Var(_) => false
+      case ElemSet(_) => true
+      case Compl(t) => t.ground
+      case Inter(_, cstsPos, varsPos, _, cstsNeg, varsNeg, other) =>
+        cstsPos.isEmpty &&
+          varsPos.isEmpty &&
+          cstsNeg.isEmpty &&
+          varsNeg.isEmpty &&
+          other.forall(_.ground)
+      case Union(_, cstsPos, varsPos, _, cstsNeg, varsNeg, other) =>
+        cstsPos.isEmpty &&
+          varsPos.isEmpty &&
+          cstsNeg.isEmpty &&
+          varsNeg.isEmpty &&
+          other.forall(_.ground)
+    }
+
+    /** Returns all [[Var]] and [[Cst]] that occur in `this`. */
+    final def unknowns: SortedSet[Int] = this match {
+      case Univ => SortedSet.empty
+      case Empty => SortedSet.empty
+      case Cst(c) => SortedSet(c)
+      case ElemSet(_) => SortedSet.empty
+      case Var(x) => SortedSet(x)
+      case Compl(t) => t.unknowns
+      case Inter(_, cstsPos, varsPos, _, cstsNeg, varsNeg, other) =>
+        SortedSet.from(
+          cstsPos.iterator.map(_.c) ++
+            varsPos.iterator.map(_.x) ++
+            cstsNeg.iterator.map(_.c) ++
+            varsNeg.iterator.map(_.x) ++
+            other.iterator.flatMap(_.unknowns)
+        )
+      case Union(_, cstsPos, varsPos, _, cstsNeg, varsNeg, other) =>
+        SortedSet.from(
+          cstsPos.iterator.map(_.c) ++
+            varsPos.iterator.map(_.x) ++
+            cstsNeg.iterator.map(_.c) ++
+            varsNeg.iterator.map(_.x) ++
+            other.iterator.flatMap(_.unknowns)
+        )
+    }
+
+    /**
+      * Returns the number of connectives in the unary/binary representation of `this`.
+      *
+      * [[Compl]], [[Union]], and [[Inter]] are connectives.
+      */
+    final def size: Int = {
+      var workList = List(this)
+      var counter = 0
+
+      /** Updates `counter` and `workList` given intersection or union subformulas. */
+      @inline
+      def countSetFormulas(
+                            elemPos: Option[ElemSet], cstsPos: Set[Cst], varsPos: Set[Var],
+                            elemNeg: Option[ElemSet], cstsNeg: Set[Cst], varsNeg: Set[Var],
+                            other: List[SetFormula]
+                          ): Unit = {
+        val negElemSize = elemNeg.size
+        val negCstsSize = cstsNeg.size
+        val negVarsSize = varsNeg.size
+        val subformulas = elemPos.size + cstsPos.size + varsPos.size +
+          negElemSize + negCstsSize + negVarsSize + other.size
+        val connectives = negElemSize + negCstsSize + negVarsSize
+        counter += (subformulas - 1) + connectives
+        workList = other ++ workList
+      }
+
+      while (workList.nonEmpty) {
```

Comment:
Why is this a fixpoint? I mean its okay, but could have been a recursive definition?

Reply:
It can't be tail-recursive if it is a recursive function, so I did this instead

Reply:
its a bit more complex but its also self-contained so this complexity doesn't influence other functions etc.

Reply:
But again, does it have to be tail recursive?

Reply:
i like it because
- faster: this is a bottle neck of performance so everything counts
- low maintenance cost: its selfcontained
- better errors: its tailrecursive, so it allows us to give TooComplex more often instead of StackOverflow
- low effort: its already written

If anything is wrong/weird its easy to go back to recursive def. I think this is better

Reply:
Then add those comments.